### ISSUE TITLE
eslint do not work

### ISSUE THREAD

**Issue by Amantel** [drmargarido/linters#22]
I have: 
* cloned linter repo to /data/plugins. 
* copied eslinter to the same folder
* restarted lite
* added some stuff that should raise warnings and errors to the .js file
* saved file
* got 'linter found no warnings' message

**Comment by drmargarido** [comment]
Hello, thanks for the report.
Can you send me the file that you are trying to lint?

**Comment by Amantel** [comment]
Sure. There is a lot of info that is not required, but why not: 
```
const express = require('express');
const _ = require('lodash');

const General = require('../models/general');
const TL = require('../helpers/async');
const Tech = require('../helpers/technical');

// router
const router = express.Router();

router.get('/home', async (req, res) => res.render('home.ejs', {
  ** *** *   currentURL: req.originalUrl
}));

function findAllChildrenForCategory(categoryId, shopInfo) {
  let allChildren = [];
  function findCategories(id) {
    const firstLevelChildren = shopInfo.categories.filter((cat) => cat.parentId === id);
    allChildren = allChildren.concat(firstLevelChildren);
    firstLevelChildren.forEach((cat) => findCategories(cat.id));
    return true;
  }
  findCategories(categoryId);
  return allChildren;
}

function composeCategoryIds(id, shopInfo) {
  const children = findAllChildrenForCategory(id, shopInfo).map((cat) => cat.id);
  return children.concat(id);
}

router.get('/home/products/:shopId/byCategories', async (req, res) => {
  const shopId = req.params.shopId || null;

  const [errGetShop, shopInfo] = await TL.to(General.getShopCategories(shopId));
  if (errGetShop) return TL.answerWithError(res, 500, 'DB ERROR', errGetShop);

  if (!shopInfo) return TL.answerWithError(res, 400, 'Shop not found', null);

  function getChildren(category) {
    let firstLevelChildren = shopInfo.categories.filter((cat) => cat.parentId === category.id);
    const hasChildren = firstLevelChildren.length > 0;
    if (hasChildren) {
      firstLevelChildren = firstLevelChildren.map((cat) => {
        cat.children = getChildren(cat);
        return cat;
      });
    }
    const children = firstLevelChildren;

    return children;
  }

  const rootCategories = shopInfo.categories.filter((c) => c.parentId === null);
  const categoriesTree = rootCategories.map((c) => {
    c.children = getChildren(c);
    return c;
  });

  return res.render('categories.ejs', {
    currentURL: req.originalUrl, shopId: req.params.shopId, categoriesTree, shopInfo
  });
});

router.get('/home/products/:shopId/byProducts', async (req, res) => {
  const shopId = req.params.shopId || null;
  const categoryId = req.query.categoryId || null;
  let categoriesList = [];
  const [errGetShop, shopInfo] = await TL.to(General.getShopCategories(shopId));
  if (errGetShop) return TL.answerWithError(res, 500, 'DB ERROR', errGetShop);

  if (!shopInfo) return TL.answerWithError(res, 400, 'Shop not found', null);

  if (categoryId) {
    categoriesList = composeCategoryIds(categoryId, shopInfo);
  }

  return res.render('categories.ejs', {
    currentURL: req.originalUrl, shopId: req.params.shopId, categoriesList, shopInfo
  });
});

router.get('/api/contracts', async (req, res) => {
  const offset = parseInt(req.query.offset, 10) || 0;
  const limit = Math.max(parseInt(req.query.limit, 10) || 50, 100);
  const search = req.query.search || null;
  const sort = req.query.sort || null;

  const [errTotal, contractsTotal] = await TL.to(General.getContractsTotal({}));
  const [err, contracts] = await TL.to(General.getContracts({}, offset, limit));

  if (err || errTotal) return TL.answerWithError(res, 500, 'DB ERROR', err || errTotal);

  return res.json({ items: contracts, count: contractsTotal });
});

router.get('/api/shops/:shopId/categories/tree', async (req, res) => { // TODO: should be PUT
  const shopId = req.params.shopId || null;

  // if (categoriesList.length === 0) return res.status(422).json({ error: true, error_info: 'categories list must be not empty' });

  const [errGetShop, shopInfo] = await TL.to(General.getShopCategories(shopId));
  if (errGetShop) return TL.answerWithError(res, 500, 'DB ERROR', errGetShop);

  if (!shopInfo) return TL.answerWithError(res, 400, 'Shop not found', null);

  function getChildren(category) {
    let firstLevelChildren = shopInfo.categories.filter((cat) => cat.parentId === category.id);
    const hasChildren = firstLevelChildren.length > 0;
    if (hasChildren) {
      firstLevelChildren = firstLevelChildren.map((cat) => {
        cat.children = getChildren(cat);
        return cat;
      });
    }
    const children = firstLevelChildren;

    return children;
  }

  const rootCategories = shopInfo.categories.filter((c) => c.parentId === null);
  const categoriesTree = rootCategories.map((c) => {
    c.children = getChildren(c);
    return c;
  });

  return res.json({
    categoriesTree,
    shopInfo,
  });
});

router.post('/api/shops/:shopId/categories/toggle', async (req, res) => { // TODO: should be PUT
  const categoriesList = Tech.arrayFromParam(req.body.list);
  const action = req.body.action || null;
  const shopId = req.params.shopId || null;

  if (categoriesList.length === 0) return res.status(422).json({ error: true, error_info: 'categories list must be not empty' });
  if (!action || (action !== 'block' && action !== 'allow')) return res.status(422).json({ error: true, error_info: 'action must be present with valid argument' });

  const [errGetShop, shopInfo] = await TL.to(General.getShopCategories(shopId));
  if (errGetShop) return TL.answerWithError(res, 500, 'DB ERROR', errGetShop);

  if (!shopInfo) return TL.answerWithError(res, 400, 'Shop not found', null);

  const listOfCategoriesIds = categoriesList.map((id) => composeCategoryIds(id, shopInfo));

  const plainListOfCategoriesIds = [...new Set(listOfCategoriesIds.reduce((total, amount) => total.concat(amount), []))];

  const actionObject = action === 'allow' ? { $set: { zabotaAllowed: true }, $unset: { zabotaBlocked: true } } : { $set: { zabotaBlocked: true }, $unset: { zabotaAllowed: true } };

  const [errUpdateProducts, result] = await TL.to(General
    .updateProducts({ shopId, categoryId: { $in: plainListOfCategoriesIds } }, actionObject,));
  if (errUpdateProducts) return res.status(500).json({ error: true, error_info: errUpdateProducts });

  let categoriesWhiteList = shopInfo.categoriesWhiteList || [];
  if (action === 'allow') {
    categoriesWhiteList = [...new Set(plainListOfCategoriesIds.concat(categoriesWhiteList))];
  } else {
    categoriesWhiteList = _.difference(categoriesWhiteList, plainListOfCategoriesIds);
  }

  const [errUpdateWhiteList] = await TL.to(General.updateShop(shopId, { categoriesWhiteList }));
  if (errUpdateWhiteList) return TL.answerWithError(res, 500, 'DB ERROR', errUpdateWhiteList);

  return res.json({
    plainListOfCategoriesIds,
    action,
    result: { modified: result.modifiedCount, matched: result.matchedCount }
  });
});

router.get('/api/shops/:shopId/products/toggle', async (req, res) => { // TODO: should be PUT
  const productIdsList = Tech.arrayFromParam(req.query.list);
  const action = req.query.action || null;
  const shopId = req.params.shopId || null;

  if (productIdsList.length === 0) return TL.answerWithError(res, 422, 'products list must be not empty', null);
  if (!action || (action !== 'block' && action !== 'allow')) return res.status(422).json({ error: true, error_info: 'action must be present with valid argument' });

  const actionObject = action === 'allow' ? { $set: { zabotaAllowed: true }, $unset: { zabotaBlocked: true } } : { $set: { zabotaBlocked: true }, $unset: { zabotaAllowed: true } };

  const [errUpdateProducts, result] = await TL.to(General
    .updateProducts({ shopId, id: { $in: productIdsList } }, actionObject,));
  if (errUpdateProducts) return TL.answerWithError(res, 500, 'DB ERROR', errUpdateProducts);

  return res.json({
    productIdsList,
    action,
    result: { modified: result.modifiedCount, matched: result.matchedCount }
  });
});

module.exports = router;
```

I inserted the problem on line 12 (with ***).
I expect linter to tell me that some strange stuff is happening there, but it does not.
Maybe I am missing something? Like adding settings file, or telling main linter where to look for additional linters?

**Comment by drmargarido** [comment]
I just tested and it reported well. 

![Screenshot from 2021-02-16 12-18-36](https://user-images.githubusercontent.com/16456150/108062368-9f2b2980-7051-11eb-9a94-eaef36581709.png)

In what operating system are you? Did you try to run eslint directly on the file to see if it works? eslint needs an initial configuration file that is generated with eslint --init.

**Comment by Amantel** [comment]
It works.
![image](https://user-images.githubusercontent.com/575862/108064924-5d5cac00-706e-11eb-9c10-6759a2340fca.png)
Also it works inside VS Code 
![image](https://user-images.githubusercontent.com/575862/108064999-76fdf380-706e-11eb-8ec2-e747bc60f8fa.png)


Maybe the problem is because eslint is not globally installed? I used `./node_modules/.bin/eslint ./controllers/home.js` to launch it manually.

Also I have got .eslintrc (it was hidden, so I missed it).

Here is it's content:
```
{
	"extends": "airbnb-base",
	"rules": {
		"no-tabs": 0,
		"comma-dangle": 0,
		"indent": [
			"error",
			2
		],
		"no-restricted-globals": "off",
		"no-param-reassign": "off",
		"prefer-destructuring": "off",
		"no-unused-vars": "warn",
		"no-underscore-dangle": 0,
		"object-curly-newline": 0,
		"consistent-return": "off",
		"max-len": [
			1,
			125,
			2,
			{
				"ignoreComments": true,
				"ignoreStrings": true,
				"ignoreUrls": true
			}
		],
		"no-console": [
			"error",
			{
				"allow": [
					"log",
					"warn",
					"error",
					"time",
					"timeEnd"
				]
			}
		],
		"linebreak-style": 0,
		"no-shadow": [
			"warn",
			{
				"allow": [
					"err",
					"params"
				]
			}
		],
		"no-use-before-define": [
			2,
			{
				"functions": false
			}
		]
	}
}
```

**Comment by Amantel** [comment]
My OS is Ubuntu

**Comment by drmargarido** [comment]
Yes, right now the linter_eslint assumes you have the eslinter command available in your path.

**Comment by Amantel** [comment]
Can I hack it somehow? 
For example change https://github.com/drmargarido/linters/blob/master/linter.lua#L95 from absolute path to local path, or something?
Or manually add some args to linter?

**Comment by drmargarido** [comment]
It will still not work since you have the binary inside the node_modules folder. 
So it's better to be added as a special rule in the https://github.com/drmargarido/linters/blob/master/linter_eslint.lua.
Check if eslint exists globally, if it doesn't try to use the local path in node modules.

**Comment by Amantel** [comment]
On line twelve.
It works, thank you!
